src/zh/zaimanhua/src/eu/kanade/tachiyomi/extension/zh/zaimanhua/Zaimanhua.kt

```diff
@@ -0,0 +1,212 @@
+package eu.kanade.tachiyomi.extension.zh.zaimanhua
+
+import android.app.Application
+import android.content.SharedPreferences
+import androidx.preference.EditTextPreference
+import androidx.preference.ListPreference
+import androidx.preference.PreferenceScreen
+import eu.kanade.tachiyomi.network.GET
+import eu.kanade.tachiyomi.network.POST
+import eu.kanade.tachiyomi.network.interceptor.rateLimit
+import eu.kanade.tachiyomi.source.ConfigurableSource
+import eu.kanade.tachiyomi.source.model.FilterList
+import eu.kanade.tachiyomi.source.model.MangasPage
+import eu.kanade.tachiyomi.source.model.Page
+import eu.kanade.tachiyomi.source.model.SChapter
+import eu.kanade.tachiyomi.source.model.SManga
+import eu.kanade.tachiyomi.source.online.HttpSource
+import okhttp3.FormBody
+import okhttp3.Headers
+import okhttp3.HttpUrl
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.Interceptor
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.RequestBody
+import okhttp3.Response
+import uy.kohesive.injekt.Injekt
+import uy.kohesive.injekt.api.get
+import java.security.MessageDigest
+
+class Zaimanhua : HttpSource(), ConfigurableSource {
+    override val lang = "zh"
+    override val supportsLatest = true
+
+    override val name = "再漫画"
+    override val baseUrl = "https://manhua.zaimanhua.com"
+    private val apiUrl = "https://v4api.zaimanhua.com/app/v1"
+    private val accountApiUrl = "https://account-api.zaimanhua.com/v1"
+
+    private val preferences: SharedPreferences =
+        Injekt.get<Application>().getSharedPreferences("source_$id", 0x0000)
+
+    override val client: OkHttpClient =
+        network.client.newBuilder().rateLimit(5).addInterceptor(::authIntercept).build()
+
+    private fun authIntercept(chain: Interceptor.Chain): Response {
+        val request = chain.request()
+        if (request.url.host != "v4api.zaimanhua.com" || !request.headers["authorization"].isNullOrBlank()) {
+            return chain.proceed(request)
+        }
+
+        var token: String = preferences.getString("TOKEN", "")!!
+        if (token.isBlank() || !isValid(token)) {
+            val username = preferences.getString("USERNAME", "")!!
+            val password = preferences.getString("PASSWORD", "")!!
+            token = getToken(username, password)
+            if (token.isBlank()) {
+                preferences.edit().putString("TOKEN", "").apply()
+                preferences.edit().putString("USERNAME", "").apply()
+                preferences.edit().putString("PASSWORD", "").apply()
```

Comment:
throw an exception when username or password is incorrect

Reply:
login is not required, no need to throw exception

Reply:
user should know when their credentials are incorrect

Reply:
Will the exception make this request fail? Is there another way to alert the user but not make this request fail?

I apologize for my lack of basic knowledge.

Reply:
yes it will cancel the request. no other way to show user facing error/snackbar otherwise... you can leave it as is ig...

Reply:
I decided to leave it as is.
Users will know if they are logged in by the non-editable TOKEN setting, and will know that they have failed to log in by the USERNAME/PASSWORD setting being cleared.